### PR TITLE
refactor(inverted_index): integrate puffin manager with sst indexer

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

#4246 

## What's changed and what's your intention?

...to make room for the addition of full-text indexer

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Added a new configuration parameter to control the compression of inverted index data.
  
- **Deprecations**
  - Deprecated `intermediate_path` setting in favor of `region_engine.mito.index.aux_path`.

- **Refactor**
  - Refactored index creation process, introducing new types and methods to streamline operations.

- **Bug Fixes**
  - Updated expected values in tests to align with modified region statistics behavior.

- **Documentation**
  - Updated configuration examples to reflect new and deprecated settings.

### CODE REVIEW

**Comment by zhongzc** [comment]
@coderabbitai pause

File: src/mito2/src/sst/index.rs [hunk 1]
```diff
@@ -12,159 +12,155 @@
...
+    /// Inverted index output.
+    pub inverted_index: InvertedIndexOutput,
+}
 
-const INDEX_BLOB_TYPE: &str = "greptime-inverted-index-v1";
+/// Output of the inverted index creation.
+#[derive(Debug, Clone, Default)]
+pub struct InvertedIndexOutput {
+    /// Whether the index is available.
+    pub available: bool,
```
Comment (evenyag) [inline src/mito2/src/sst/index.rs:53]: If we can use `index_size` to check whether the index is available, do we still need the `available` flag?

File: src/mito2/src/sst/index.rs [hunk 2]
```diff
@@ -12,159 +12,155 @@
...
+        self.do_update(batch).await;
 
-        if let Some(creator) = self.inner.as_ref() {
-            let memory_usage = creator.memory_usage();
-            INDEX_CREATE_MEMORY_USAGE.add(memory_usage as i64 - self.last_memory_usage as i64);
-            self.last_memory_usage = memory_usage;
-        } else {
-            INDEX_CREATE_MEMORY_USAGE.sub(self.last_memory_usage as i64);
-            self.last_memory_usage = 0;
-        }
```
Comment (evenyag) [inline src/mito2/src/sst/index.rs:76]: So this is removed?

File: src/mito2/src/sst/index/inverted_index/creator.rs [hunk 3]
```diff
@@ -228,7 +210,7 @@ impl SstIndexCreator {
             // non-null value -> Some(encoded_bytes), null value -> None
             let value = value.is_some().then_some(self.value_buf.as_slice());
             self.index_creator
-                .push_with_name_n(column_id, value, n)
+                .push_with_name_n(&column_id.to_string(), value, n)
```
Comment (evenyag) [inline src/mito2/src/sst/index/inverted_index/creator.rs:213]: We have to create a column id string each time. Can we avoid this?

File: src/mito2/src/sst/index.rs (hunk 2 above)
Comment (zhongzc) [reply src/mito2/src/sst/index.rs:76]: If `creator` is `None`, The `0` memory usage acts the same for two branches.

File: src/mito2/src/compaction/compactor.rs [hunk 4]
```diff
@@ -311,7 +311,7 @@ impl Compactor for DefaultCompactor {
                         file_size: sst_info.file_size,
                         available_indexes: {
                             let mut indexes = SmallVec::new();
-                            if sst_info.index_metadata.inverted_index.available {
+                            if sst_info.index_metadata.inverted_index.index_size > 0 {
```
Comment (evenyag) [inline src/mito2/src/compaction/compactor.rs:314]: non-blocking: We can add a `is_available()` method that checks the index size.
```suggestion
                            if sst_info.index_metadata.inverted_index.is_available() {
```

**Review by evenyag — APPROVED** [review]

File: src/mito2/src/sst/index/indexer/finish.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+use crate::sst::index::puffin_manager::SstPuffinWriter;
+use crate::sst::index::statistics::{ByteCount, RowCount};
+use crate::sst::index::{IndexOutput, Indexer, InvertedIndexOutput};
+
+impl Indexer {
+    pub(crate) async fn do_finish(&mut self) -> IndexOutput {
+        let mut output = IndexOutput::default();
+
+        let Some(mut writer) = self.puffin_writer.take() else {
+            return IndexOutput::default();
```
Comment (v0y4g3r) [inline src/mito2/src/sst/index/indexer/finish.rs:28]:
```suggestion
            return output;
```

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**